### PR TITLE
Wait for extension unlock before processing eth_requestAccounts

### PR DESCRIPTION
**Pull request by rekmarks**
### Summary

- Adds `unlock` event and related functionality to controllers:
  - Updates `eth-keyring-controller` to version with `unlock` event
  - `MetamaskController`
    - Add `addUnlockListener` method
  - `AppStateController`
    - Add functionality for waiting on extension unlock
    - Add methods: `getUnlockPromise`, `handleUnlock`
  - Adds to notification badge count for every request waiting on unlock
- `PermissionsController`: Waits for the extension to become unlocked before attempting to process `eth_requestAccounts`, _if_ the `eth_accounts` permission is already granted
  - Returns a `resourceUnavailable` Ethereum JSON RPC error for all requests following the first while waiting
    - This avoids memory leaks in case we get spammed with a bunch of requests while the extension is locked
  - If the `eth_accounts` permission is _not_ granted, this is handled by functionality introduced in #8148 

### Details

- Does _not_ open the extension popup when receiving `eth_requestAccounts` if the `eth_accounts` permission is already granted
  - We should create a separate PR if we want to support this functionality, as we have to make the popup close on unlock if there are no notifications/confirmations to attend to, which is its own set of work

### CODE REVIEW

File: app/scripts/metamask-controller.js
Comment (rekmarks) [inline app/scripts/metamask-controller.js]: Is this equivalent to that which it replaced?

Comment (Gudahtt) [reply app/scripts/metamask-controller.js]: Yep! `KeyringController` is in `this.memstore`, so it's included in the huge flattened state that `this.getState` returns.

Though I would suggest updating `this.isUnlocked` to return `this.keyringController.memStore.getState().isUnlocked` instead of calling `this.getState`, because the former is far more efficient. `getState` will ask for the state of _all_ controllers then flatten and combine them. Lots of work to get just one property.

File: app/scripts/controllers/permissions/methodMiddleware.js
Comment (Gudahtt) [inline app/scripts/controllers/permissions/methodMiddleware.js]: I'm not sure why this `catch` block is here - it should be fine to just `await` the unlock promise I would think

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [inline app/scripts/metamask-controller.js]: Regarding the badge count; maybe we could avoid the complexities you mentioned in the PR description by keeping track of these unlock promises, and counting those in the badge as well.

So instead of returning the promise here, we could add it to a queue that we'd store on the instance:

```
return new Promise((resolve) => {
  if (this.isUnlocked()) {
    resolve()
  } else {
    this.waitingForUnlock.push({ resolve })
  }
})
```

Then on unlock, we could drain the queue:

```
while (this.waitingForUnlock.length >= 0) {
  this.waitingForUnlock.shift().resolve()
}
```

The `updateBadge` function can just count this queue and add that to the badge count.
Though we'd also need to handle updates 🤔 It seems like this'd be better suited to a controller, which we have an established pattern for adding subscriptions to, but I don't know where it'd fit.

File: app/scripts/controllers/permissions/methodMiddleware.js
Comment (rekmarks) [reply app/scripts/controllers/permissions/methodMiddleware.js]: Yeah, it's totally fine. Just me being paranoid. I'm addressing this by `This Promise will never reject.` to the `getUnlockPromise` docstring.

File: app/scripts/metamask-controller.js
Comment (rekmarks) [reply app/scripts/metamask-controller.js]: Sweet, done.

File: app/scripts/metamask-controller.js
Comment (rekmarks) [reply app/scripts/metamask-controller.js]: This was implemented and added to the `AppStateController`. That class now extends `EventEmitter`, and the background subscribes to its `updateBadge` event per the usual pattern.

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [inline app/scripts/metamask-controller.js]: Is the extension unlocked in this branch? 🤔 I don't follow why that would be the case

File: app/scripts/controllers/permissions/methodMiddleware.js
Comment (Gudahtt) [inline app/scripts/controllers/permissions/methodMiddleware.js]: Is this case needed? I thought we were already preventing this in #8148

**Review by Gudahtt — CHANGES_REQUESTED** [review]
I'm not sure we're handing keyring construction properly; i.e. we might need to emit `unlocked` after the initial keyring construction, to drain anything queued up during onboarding

File: app/scripts/metamask-controller.js
Comment (rekmarks) [reply app/scripts/metamask-controller.js]: I think it is, if you look at the statement starting on L1776 above:
```javascript
if (!addresses.length) {
  return
}
```

Comment (Gudahtt) [reply app/scripts/metamask-controller.js]: The keyring can still have addresses _and_ be locked though

File: app/scripts/controllers/permissions/methodMiddleware.js
Comment (rekmarks) [reply app/scripts/controllers/permissions/methodMiddleware.js]: Yeah, it's needed for when the extension is locked _and_ the `eth_accounts` permission is already granted.

If the permission is _not_ granted, the permission will be requested a pending approval will be created. If an approval is already pending, any additional requests will be rejected. This is what #8148 does.

Without this PR, if the permission _is_ granted _and_ the extension is locked, this method breaks. Why? Because `let accounts = await getAccounts()` will return `[]`, and then create a pending approval. This will re-request the permission completely, even though it's already granted, and should just return the already-approved accounts.

In short, without this PR, this method can't tell whether `getAccounts()` returns `[]` because the extension is locked or because the permission isn't granted. This PR allows this method to distinguish between these cases and behave appropriately in both.

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [reply app/scripts/metamask-controller.js]: Or at least I think it can? Could be wrong. I looked at the KeyringController and I couldn't see how the presence of addresses would guarantee it was unlocked.

File: app/scripts/controllers/permissions/methodMiddleware.js
Comment (Gudahtt) [reply app/scripts/controllers/permissions/methodMiddleware.js]: Right - I just meant to question whether this `if (isProcessingRequestAccounts) {` branch was needed. I had thought the `await getUnlockedPromise()` below would solve the problem you're describing.

Though I guess #8148 wouldn't stop these unlock requests from piling up either, so maybe this second-layer of preventing duplicate requests is necessary 🤔

Comment (rekmarks) [reply app/scripts/controllers/permissions/methodMiddleware.js]: Ah, I see.

And yeah, that statement is exactly to prevent unlock requests from piling up.

File: app/scripts/metamask-controller.js
Comment (rekmarks) [reply app/scripts/metamask-controller.js]: This has been fixed by updating `eth-keyring-controller`, which now emits an `unlock` event every time it becomes unlocked.

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [inline app/scripts/metamask-controller.js]: Could `this._onUnlock` be called directly here rather than re-emitting it on the controller itself? It can be a bit confusing tracking all of the events going around so if we can minimize those it would be nice.

Comment (Gudahtt) [reply app/scripts/metamask-controller.js]: Oh I see - it was done here because `keyringController` is constructed after the app state controller. 🤔

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [inline app/scripts/metamask-controller.js]: I'm not sure about this stuff being moved. The `wasLocked` variable implies that this only ran in the case where the keyring controller _was_ locked but is no longer, but... I don't understand how that could be the case. The `state` passed in is the _current_ state, and it's checking that it's _not_ unlocked, so... this doesn't seem equivalent?

File: app/scripts/metamask-controller.js
Comment (rekmarks) [reply app/scripts/metamask-controller.js]: Yeah, I think at least one of the commits with failed tests is from me trying to do otherwise.

File: app/scripts/metamask-controller.js
Comment (rekmarks) [reply app/scripts/metamask-controller.js]: >>As part of some permissions-related functionality, I need to reliably emit an event every time the extension becomes unlocked

>This should be possible.

>> I understand that `addresses.length === 0 if` the extension/keyring controller is locked (I understand those to be one and the same), but it’s unclear to me why `addresses.length > 0 && !isUnlocked` would ever be `true`

>Looking at the KC code, only only time an 'update' event is emitted with isUnlocked set to false is after a call to #setLocked, which also unsets/empties the keyrings. So, yes, `addresses.length > 0 && !isUnlocked` would never be true, with the current state of the code. (https://github.com/MetaMask/KeyringController/blob/5a95355fa4c0856c0f8d7c0cb57fa6af91364482/index.js#L114-L126)

>BUT

>No, that's wrong. All calls to updateState would also emit 'update' events, so lines 121 and 124 will each emit an `'update'` event (https://github.com/MetaMask/KeyringController/blob/5a95355fa4c0856c0f8d7c0cb57fa6af91364482/index.js#L121-L124) the first with `isUnlocked === false && keyrings.length > 0`, the second with `isUnlocked === false && keyrings.length === 0`

Comment (rekmarks) [reply app/scripts/metamask-controller.js]: All of which is to say, I think it is equivalent?

Comment (Gudahtt) [reply app/scripts/metamask-controller.js]: I don't quite follow. That makes me thing that this block was executed _on lock_, not on _unlock_.

Comment (rekmarks) [reply app/scripts/metamask-controller.js]: Hm. On review, I agree, but I _think_ the attention was the opposite?
- https://github.com/MetaMask/metamask-extension/pull/5066
- The commit that added the `wasLocked` block: https://github.com/MetaMask/metamask-extension/pull/5066/commits/003d445a98164dac0c0529dfd69f5e3987d7d05c
- The commit that added the tests: https://github.com/MetaMask/metamask-extension/pull/5066/commits/e803b8e047b5acad9143fe28d99d9e7d65211f46

If the extension is locked, we'll see `keyrings === []`, and there will be no accounts to select. I interpreted this code as trying to cover for the case where a keyring had been removed since the `wasLocked` (now `_onUnlock`) block was executed.

That can still happen with a `keyringController` state update, I suppose, but I don't see how it would happen while the keyring controller/extension is locked.

I'm seeing if e2e tests pass if the block is deleted.

File: yarn.lock
Comment (Gudahtt) [inline yarn.lock]: Could you update this one too? This is the one GABA is using, so it'd be ideal to keep it in sync

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [reply app/scripts/metamask-controller.js]: Yeah I think you're right. I don't understand why this would be on-lock either.
@whymarrh do you recall?

File: yarn.lock
Comment (rekmarks) [reply yarn.lock]: Fixed in 56433c5

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [reply app/scripts/metamask-controller.js]: Actually it looks like this is covered already by the `this.preferencesController.syncAddresses(accounts)`call in `submitPassword`. That ensures the accounts are valid upon unlock. I don't think we need this stuff after all.

File: yarn.lock
Comment (Gudahtt) [reply yarn.lock]: That didn't seem to work - GABA is still using v5.5.0 of the keyring controller

Comment (rekmarks) [reply yarn.lock]: Hm, we should update that.

File: app/scripts/metamask-controller.js
Comment (rekmarks) [reply app/scripts/metamask-controller.js]: Resolved via #8224

File: yarn.lock
Comment (rekmarks) [inline yarn.lock]: Ugh, is this okay?

Comment (Gudahtt) [reply yarn.lock]: This happened locally for me as well!

I think this change _should_ have been in #8206 , but for some reason it wasn't. Sometimes Yarn gets into a state where it won't notice there is redundancy in the lockfile - I don't quite understand why.

So I think this is fine 👍

**Review by Gudahtt — APPROVED** [review]
LGTM! Seems to work for me locally after that last Keyring Controller update

**Comment by rekmarks** [comment]
Note for posterity: force-merged because a codeowner temporarily lost access to GitHub.

**Pull request merged**
